src/main/java/org/breedinginsight/daos/UserDAO.java

```diff
@@ -83,36 +84,77 @@ public UserDAO(Configuration config, DSLContext dsl) {
         return dsl.select()
                 .from(BI_USER)
                 .leftJoin(SYSTEM_USER_ROLE).on(BI_USER.ID.eq(SYSTEM_USER_ROLE.BI_USER_ID))
-                .leftJoin(SYSTEM_ROLE).on(SYSTEM_USER_ROLE.SYSTEM_ROLE_ID.eq(SYSTEM_ROLE.ID));
+                .leftJoin(SYSTEM_ROLE).on(SYSTEM_USER_ROLE.SYSTEM_ROLE_ID.eq(SYSTEM_ROLE.ID))
+                .leftJoin(PROGRAM_USER_ROLE).on(PROGRAM_USER_ROLE.USER_ID.eq(BI_USER.ID))
+                .leftJoin(PROGRAM).on(PROGRAM_USER_ROLE.PROGRAM_ID.eq(PROGRAM.ID))
+                .leftJoin(ROLE).on(PROGRAM_USER_ROLE.ROLE_ID.eq(ROLE.ID))
+                    .and(PROGRAM.ACTIVE.eq(true));
+
     }
 
     private List<User> parseRecords(List<Record> records) {
 
-        List<User> resultUsers = new ArrayList<>();
+        Map<String, User> userMap = new HashMap<>();
 
         // Parse the result
         for (Record record : records) {
             // program user exists
             User userRecord = User.parseSQLRecord(record);
             SystemRole systemRole = SystemRole.parseSQLRecord(record);
-            Optional<User> existingUser = resultUsers.stream()
-                    .filter(p -> p.getId().equals(userRecord.getId()))
-                    .findFirst();
-
-            User user;
-            if (!existingUser.isPresent()) {
-                resultUsers.add(userRecord);
-                user = userRecord;
+            ProgramUser programUser = ProgramUser.parseSQLRecord(record);
+            Role programRole = Role.parseSQLRecord(record);
+            Program program = Program.parseSQLRecord(record);
+
+            User existingUser;
+            if (userMap.containsKey(userRecord.getId().toString())) {
+                existingUser = userMap.get(userRecord.getId().toString());
             } else {
-                user = existingUser.get();
+                userMap.put(userRecord.getId().toString(), userRecord);
+                existingUser = userRecord;
             }
 
+            // Add our system role
             if (systemRole.getDomain() != null) {
-                user.addRole(systemRole);
+                Optional<SystemRole> systemRoleExists = Utilities.findInList(existingUser.getSystemRoles(),
+                        systemRole, SystemRole::getId);
+                if (!systemRoleExists.isPresent()){
+                    existingUser.addRole(systemRole);
+                }
             }
+
+            // Add our program
+            if (programUser.getProgramId() != null){
+                Optional<ProgramUser> programUserExists = Utilities.findInList(existingUser.getProgramRoles(),
```

Comment:
When the program is deactivated a program role is returned but without the roles array which I assume is what you want? I didn't see a test case for this scenario.

Reply:
I'll have to fix this. The intended functionality is for the program role to not be returned if the program is inactive.

Reply:
I pushed up a fix for this so if the program is inactive, the program role is not returned. I had to add onto our Jooq parsing pattern because our previous flow did not support inner select queries, which I needed for this case. 

I ended up needing to generate field aliases for the fields of the inner select, and creating another `parseSqlRecord` method on the models. I think that as far as the query construction goes, the field alias generation with the table alias is the best I could come up with. It kind of mimics the behavior you need to follow when writing raw sql. I wish there was a way to combine the meat of the logic in the `parseSqlRecord` into one share function, but our normal `parseSqlRecord(record)` uses a Field as the lookup, and our `parseSqlRecord(record, alias)` method uses a string, which just looks at the column name. 

Any thoughts on this implementation would be appreciated! @timparsons @dmeidlin You might also be interested in checking this out since it is a base infrastructure change that will affect a lot of code in the future.

Reply:
I had some other comments written up on ideas for `parseSqlRecord` in particular, however, as I got looking more at this code, I think we are reaching the point of over-complicating a SQL query.  I think it's worth re-visiting the discussion of when to have a single query for all the data, and when to look at splitting it up into two calls.

IMO, I would split out the retrieval of a users programs and roles to a second SQL call.  I feel it will make the code cleaner, and easier to follow, as well as allowing the use of the original `parseSqlRecord` method.

Reply:
Yep, I agree that in this case two calls is cleaner! In the future, I think we will still need to consider each complex query on a case by case basis to avoid slow calls and so that we don't end up doing manual joins for the database. I'm thinking of the phenotyping/genotyping data in particular when we are joining potentially millions of rows. 

Anyway, for this case, two calls doesn't add a lot of overhead so I think we are fine performance wise. I'll have the new code up soon. Thanks for the suggestion!

Reply:
Changed and pushed. 

Also, updated the Program User endpoints and the api docs to return the program as well (they didn't before). To exclude the program from the endpoints, we will have to modify the way we are using jackson, I can describe a possible solution if anyway is interested. I had a conversation with Nick about this and one of the possible solutions we came up with is to set a standard that a response referencing a nested class should not have a higher level of detail than the nested classes own response. This forces a consistency with our responses. If anyone has further thoughts on this, I am open to discussing them!